dali/c_api_2/managed_handle.h

```diff
@@ -0,0 +1,124 @@
+    // Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_C_API_2_MANAGED_HANDLE_H_
+#define DALI_C_API_2_MANAGED_HANDLE_H_
+
+#define DALI_ALLOW_NEW_C_API
+#include "dali/dali.h"
+#include "dali/core/unique_handle.h"
+
+namespace dali::c_api {
+
+template <typename HandleType, typename Actual>
+class RefCountedHandle {
+ public:
+  using handle_type = HandleType;
+  static constexpr handle_type null_handle() { return 0; }
+
+  constexpr RefCountedHandle() : handle_(Actual::null_handle()) {}
+  constexpr explicit RefCountedHandle(handle_type h) : handle_(h) {}
+  ~RefCountedHandle() { reset(); }
+
+  RefCountedHandle(const RefCountedHandle &h) {
+    handle_ = h.handle_;
+    if (*this)
+        Actual::IncRef(handle_);
+  }
+
+  RefCountedHandle(RefCountedHandle &&h) noexcept {
+    handle_ = h.handle_;
+    h.handle_ = Actual::null_handle();
+  }
+
+  RefCountedHandle &operator=(const RefCountedHandle &other) {
+    if (other.handle_) {
```

Comment:
Shouldn't you check for the &other == this first? If one does `a = a`, the L50 assignment would assign null handle after the reset() in L49.

Reply:
Nice catch!

Reply:
Actually, it's even better to check for handle equality - if they're equal, we can skip the unnecessary incref/decref and make it no-op.